typing/typecore.ml

```diff
@@ -990,13 +968,110 @@ let rec has_literal_pattern p = match p.ppat_desc with
   | Ppat_or (p, q) ->
      has_literal_pattern p || has_literal_pattern q
 
-exception Need_backtrack
-
 let check_scope_escape loc env level ty =
   try Ctype.check_scope_escape env level ty
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
+type pattern_checking_mode =
+  | Normal
+  (** We are checking user code. *)
+
+  | Counter_example of counter_example_checking_info
+(** In [Counter_example] mode, we are checking a counter-example candidate
+    produced by Parmatch. This is a syntactic pattern that represents a set of
+    values by using or-patterns (p_1 | ... | p_n) to enumerate all alternatives
+    in the counter-example search.
+
+    Parmatch does not use type information, so this pattern may exhibit two
+    issues:
+    - some parts of the pattern may be ill-typed due to GADTs.
+    - some wildcard patterns may not match any values: their type are empty.
```

Comment:
Actually these two issues are not orthogonal.
Namely, when explode a wildcard we do not so much expect it to be empty, rather we expect that all the branches will result in equations that make the remainder of the pattern ill-typed.

Reply:
In the context of this search program, "being well-typed" may have either a local or a global meaning. Local: it's well-typed as a program fragment in isolation. Global: it's well-typed inside the rest of the pattern. Because we are doing backtracking, the global meaning should be refined: there exists a typed form for the fragment such that invoking the search continuation with this form eventually succeeds. In the same style, "empty" here means that there is no well-typed value, globally: there is no value such that invoking the search continuation eventually succeeds.

Reply:
(I don't think it's worth trying to express this nuance in this part of the documentation.)